File: Interseccion.java
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class Interseccion {
    private int row, col;
    private float weight; // atributo extra (puede representar costo de permanecer en la intersección)
    private ArrayList<Arista> vecinos;
    private Semaforo semaforo = new Semaforo();

    public Interseccion(int r, int c) {
        this.row = r;
        this.col = c;
        this.weight = 0f;
        this.vecinos = new ArrayList<Arista>();
    }

    @Override
    public String toString() {
        return "(" + this.row + "," + this.col + ")";
    }

    // Calcula vecinos (misma lógica original), retorna array {vrow, vcol, hrow, hcol}
    public int[] CalcularVecinos() {
        // Vecino vertical
        int vrow = (this.col % 2 == 1) ? this.row - 1 : this.row + 1;
        int vcol = this.col;
        if (vrow < 1 || vrow > Main.maxR || vcol < 1 || vcol > Main.maxC) {
            vrow = 0;
            vcol = 0;
        }

        // Vecino horizontal
        int hrow = this.row;
        int hcol = (this.row % 2 == 1) ? this.col + 1 : this.col - 1;
        if (hrow < 1 || hrow > Main.maxR || hcol < 1 || hcol > Main.maxC) {
            hrow = 0;
            hcol = 0;
        }

        return new int[]{vrow, vcol, hrow, hcol};
    }

    // Antes: anadirvecinos(Interseccion to) que instanciaba Arista "a la loca".
    // Ahora: Controller crea la Arista adecuada (Calle/Avenida) y la añade aquí.
    public void anadirArista(Arista ar) {
        if (ar != null) {
            vecinos.add(ar);
        }
    }

    public List<Arista> getVecinos() {
        return this.vecinos;
    }

    public int getRow() {
        return row;
    }

    public int getCol() {
        return col;
    }

    public String getId() {
        return row + "," + col;
    }

    public void setWeight(float w) {
        this.weight = w;
    }

    public float getWeight() {
        return this.weight;
    }

    public Semaforo getSemaforo() { return this.semaforo; }
    public void setSemaforo(Semaforo s) { this.semaforo = s; }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Interseccion)) return false;
        Interseccion that = (Interseccion) o;
        return row == that.row && col == that.col;
    }

    @Override
    public int hashCode() {
        return Objects.hash(row, col);
    }
}

File: Interseccion.js
/**
 * Clase Interseccion - Representa una intersección en el sistema de tráfico
 * Equivalente a Interseccion.java
 */
export class Interseccion {
    constructor(row, col, gridSize = 6) {
        this.row = row;
        this.col = col;
        this.gridSize = gridSize;
        this.weight = 0;
        this.vecinos = []; // Array de Arista
    }

    /**
     * Calcula los vecinos de esta intersección según el patrón hexagonal
     * Replica la lógica de CalcularVecinos() en Java
     */
    calcularVecinos() {
        this.vecinos = [];

        // Vecino vertical (calle)
        const vrow = (this.col % 2 === 1) ? this.row - 1 : this.row + 1;
        const vcol = this.col;

        // Vecino horizontal (avenida)
        const hrow = this.row;
        const hcol = (this.row % 2 === 1) ? this.col + 1 : this.col - 1;

        return {
            vertical: { row: vrow, col: vcol, valid: this.isValidPosition(vrow, vcol) },
            horizontal: { row: hrow, col: hcol, valid: this.isValidPosition(hrow, hcol) }
        };
    }

    /**
     * Verifica si una posición es válida en la grilla
     */
    isValidPosition(row, col) {
        return row >= 1 && row <= this.gridSize && col >= 1 && col <= this.gridSize;
    }

    /**
     * Agrega un vecino (arista) a esta intersección
     */
    agregarVecino(arista) {
        this.vecinos.push(arista);
    }

    /**
     * Obtiene el identificador único de la intersección
     */
    getId() {
        return `${this.row},${this.col}`;
    }

    /**
     * Representación en string de la intersección
     */
    toString() {
        return `Interseccion(${this.row}, ${this.col})`;
    }

    /**
     * Compara si dos intersecciones son iguales
     */
    equals(other) {
        if (!other) return false;
        return this.row === other.row && this.col === other.col;
    }

    /**
     * Obtiene el peso total considerando todas las aristas vecinas
     */
    calcularPesoTotal() {
        if (this.vecinos.length === 0) return 0;
        
        const sumaWeights = this.vecinos.reduce((sum, arista) => sum + arista.weight, 0);
        return sumaWeights / this.vecinos.length;
    }
}